File: backend/app/schemas/query.py
from pydantic import BaseModel, Field
from typing import Optional, Any
from datetime import datetime


class NL2SQLRequest(BaseModel):
    datasource_id: int
    question: str = Field(..., min_length=1)
    conversation_history: Optional[list[dict]] = None


class NL2SQLResponse(BaseModel):
    sql: Optional[str]
    explanation: str
    clarification: Optional[str]
    chart_suggestion: str
    confidence: float


class ExecuteQueryRequest(BaseModel):
    datasource_id: int
    sql: str
    natural_language: Optional[str] = None
    chart_type: Optional[str] = None
    save_history: bool = True


class ExecuteQueryResponse(BaseModel):
    success: bool
    error: Optional[str]
    columns: list[str]
    rows: list[dict]
    row_count: int
    execution_time_ms: int
    ai_summary: Optional[str] = None
    chart_type: Optional[str] = None
    history_id: Optional[int] = None


class QueryHistoryResponse(BaseModel):
    id: int
    datasource_id: Optional[int]
    natural_language: str
    generated_sql: Optional[str]
    executed_sql: Optional[str]
    row_count: Optional[int]
    execution_time_ms: Optional[int]
    status: str
    error_message: Optional[str]
    ai_summary: Optional[str]
    chart_type: Optional[str]
    is_favorite: int
    created_at: Optional[datetime]

    class Config:
        from_attributes = True


class AnalyzeRequest(BaseModel):
    question: str
    sql: str
    columns: list[str]
    rows: list[dict]


class ExportRequest(BaseModel):
    history_id: int
    format: str = Field(..., pattern="^(excel|csv)$")  # excel 或 csv


class QueryFavoriteResponse(BaseModel):
    id: int
    datasource_id: Optional[int]
    natural_language: str
    generated_sql: Optional[str]
    tags: Optional[str]
    description: Optional[str]
    execute_count: int
    last_executed_at: Optional[datetime]
    created_at: Optional[datetime]
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True


class CreateFavoriteRequest(BaseModel):
    datasource_id: int
    natural_language: str
    generated_sql: Optional[str] = None
    tags: Optional[str] = None
    description: Optional[str] = None


class UpdateFavoriteRequest(BaseModel):
    tags: Optional[str] = None
    description: Optional[str] = None
    natural_language: Optional[str] = None
    generated_sql: Optional[str] = None


class SearchFavoritesRequest(BaseModel):
    keyword: Optional[str] = None
    tags: Optional[list[str]] = None
    datasource_id: Optional[int] = None
    limit: int = 50
    offset: int = 0

File: frontend/src/api/query.ts
import apiClient from './client'

export interface NL2SQLRequest {
  datasource_id: number
  question: string
  conversation_history?: Array<{ nl: string; sql?: string }>
}

export interface NL2SQLResponse {
  sql: string | null
  explanation: string
  clarification: string | null
  chart_suggestion: string
  confidence: number
}

export interface ExecuteQueryRequest {
  datasource_id: number
  sql: string
  natural_language?: string
  chart_type?: string
  save_history?: boolean
}

export interface ExecuteQueryResponse {
  success: boolean
  error: string | null
  columns: string[]
  rows: Record<string, unknown>[]
  row_count: number
  execution_time_ms: number
  ai_summary: string | null
  chart_type: string | null
  history_id: number | null
}

export interface QueryHistory {
  id: number
  datasource_id: number | null
  natural_language: string
  generated_sql: string | null
  executed_sql: string | null
  row_count: number | null
  execution_time_ms: number | null
  status: string
  error_message: string | null
  ai_summary: string | null
  chart_type: string | null
  is_favorite: number
  created_at: string | null
}

export const queryApi = {
  nl2sql: (data: NL2SQLRequest) =>
    apiClient.post<NL2SQLResponse>('/api/query/nl2sql', data).then(r => r.data),
  execute: (data: ExecuteQueryRequest) =>
    apiClient.post<ExecuteQueryResponse>('/api/query/execute', data).then(r => r.data),
  getHistory: (limit = 50, offset = 0) =>
    apiClient.get<QueryHistory[]>('/api/query/history', { params: { limit, offset } }).then(r => r.data),
  toggleFavorite: (id: number) =>
    apiClient.put<{ is_favorite: number }>(`/api/query/history/${id}/favorite`).then(r => r.data),
  export: async (historyId: number, format: 'excel' | 'csv') => {
    const response = await apiClient.post(
      '/api/query/export',
      { history_id: historyId, format },
      { responseType: 'blob' }
    )
    return response.data
  },
  createFavorite: (data: { datasource_id: number; natural_language: string; generated_sql?: string; tags?: string; description?: string }) =>
    apiClient.post('/api/query/favorites', data).then(r => r.data),
  searchFavorites: (keyword?: string, tags?: string, datasource_id?: number, limit = 50, offset = 0) =>
    apiClient.get('/api/query/favorites', { params: { keyword, tags, datasource_id, limit, offset } }).then(r => r.data),
  updateFavorite: (id: number, data: { tags?: string; description?: string; natural_language?: string; generated_sql?: string }) =>
    apiClient.put(`/api/query/favorites/${id}`, data).then(r => r.data),
  deleteFavorite: (id: number) =>
    apiClient.delete(`/api/query/favorites/${id}`).then(r => r.data),
  executeFavorite: (id: number) =>
    apiClient.post<ExecuteQueryResponse>(`/api/query/favorites/${id}/execute`).then(r => r.data),
  getFavoritesList: (datasource_id?: number, limit = 20) =>
    queryApi.searchFavorites(undefined, undefined, datasource_id, limit, 0),
}
